src/dbnode/storage/shard.go

```diff
@@ -2660,74 +2661,217 @@ func (s *dbShard) Repair(
+	defer func() {
+		if !closed {
+			if err := readerIter.Close(); err != nil {
+				// NB: log the error on ungraceful exit.
+				s.logger.Error("could not close read iterator on error", zap.Error(err))
+			}
+		}
+	}()
+
+	encoder := s.opts.EncoderPool().Get()
+	defer encoder.Close()
+	encoder.Reset(opts.Start, 0, targetSchemaDescr)
+
+	latestTargetVolume, err := s.LatestVolume(opts.Start)
+	nextVersion := latestTargetVolume + 1
+	if err != nil {
+		s.logger.Error("latestTargetVolume", zap.Error(err))
+		nextVersion = 0
+	}
+
+	writer, err := fs.NewStreamingWriter(
+		fs.StreamingWriterOptions{
+			NamespaceID:         s.namespace.ID(),
+			ShardID:             s.ID(),
+			Options:             s.opts.CommitLogOptions().FilesystemOptions(),
+			BlockStart:          opts.Start,
+			BlockSize:           s.namespace.Options().RetentionOptions().BlockSize(),
+			VolumeIndex:         nextVersion,
+			PlannedRecordsCount: uint(maxEntries),
 		},
-		FileSetType: persist.FileSetFlushType,
-		//TODO add after streaming supported - OrderByIndex: true
+	)
+	if err != nil {
+		return 0, err
 	}
-	if err := reader.Open(openOpts); err != nil {
+
+	if err := writer.Open(); err != nil {
 		return 0, err
 	}
-	defer reader.Close()
 
-	encodingOpts := encoding.NewOptions().SetBytesPool(s.opts.BytesPool())
-	bytesReader := bytes.NewReader(nil)
-	dataPointIter := m3tsz.NewReaderIterator(bytesReader, m3tsz.DefaultIntOptimizationEnabled, encodingOpts)
-	var lastWriteError error
-	var processedBlockCount int64
+	var (
+		processedTileCount  atomic.Int64
+		multiErr            xerrors.MultiError
+		dp                  ts.Datapoint
+		segmentCapacity     int
+		writerData          = make([][]byte, 2)
+	)
+
+	for readerIter.Next() {
+		seriesIter, id, encodedTags := readerIter.Current()
 
-	for {
-		id, tags, data, _, err := reader.Read()
-		if err == io.EOF {
-			break
+		for seriesIter.Next() {
+			frame := seriesIter.Current()
+			unit, singleUnit := frame.Units().SingleValue()
+			annotation, singleAnnotation := frame.Annotations().SingleValue()
+
+			if frameValues := frame.Values(); len(frameValues) > 0 {
+				lastIdx := len(frameValues) - 1
+
+				value := frameValues[lastIdx]
+				timestamp := frame.Timestamps()[lastIdx]
+
+				if !singleUnit {
+					// TODO: what happens if unit has changed mid-tile?
+					// Write early and then do the remaining values separately?
+					if lastIdx < len(frame.Units().Values()) {
+						unit = frame.Units().Values()[lastIdx]
+					}
+				}
+
+				if !singleAnnotation {
+					// TODO: what happens if annotation has changed mid-tile?
+					// Write early and then do the remaining values separately?
+					if lastIdx < len(frame.Annotations().Values()) {
+						annotation = frame.Annotations().Values()[lastIdx]
+					}
+				}
+
+				dp.Timestamp = timestamp
+				dp.Value = value
+				err = encoder.Encode(dp, unit, annotation)
+				if err != nil {
+					s.metrics.largeTilesWriteErrors.Inc(1)
+					multiErr = multiErr.Add(err)
```

Comment:
Probably also need a variable to be set, we should break out of aggregating this single series altogether if a single encode fails (since now you only have partial results).

Reply:
:+1: